### ISSUE TITLE
Implement secure temporary file handling

### ISSUE THREAD

**Issue by TobyGiacometti** [Fizzadar/pyinfra#1085]
## Is your feature request related to a problem? Please describe

I have noticed that the temporary files created by pyinfra are world-readable and often have a predictable name. For example, when using `pyinfra.operations.server.script()`, the uploaded script is stored in the temporary directory with permission bits set to 755. This has security implications if the script contains sensitive information. Another example would be `pyinfra.operations.apt.deb()` which downloads data and stores it in a temporary file with a predictable name (SHA-1 hash of the download URL). In this example, a malicious actor could simply place a Debian package with the same name in the temporary directory, and because the internal call to `pyinfra.operations.files.download()` does not use the `force` parameter, the malicious actor's Debian package would be installed.

## Describe the solution you'd like

In my opinion, the most robust solution would be to store all temporary deploy files in a secure temporary directory managed by pyinfra. I have addressed the issue in pyinfra 3.4 using a config file (`config.py`) that looks as follows:

```python
"""The pyinfra config file."""

import os
import shlex
import sys
import typing
import uuid

import pyinfra
import pyinfra.api.host
import pyinfra.connectors.ssh
import pyinfra.facts.server
import pyinfra.operations.python


def pyinfra_temp_path(
    self: pyinfra.api.host.Host, *_: typing.Any, temp_directory: str | None = None
) -> str:
    """Override for :meth:`pyinfra.api.host.Host.get_temp_filename`.

    pyinfra creates predictably named world-readable temporary files.
    This override introduces more secure handling.

    https://github.com/pyinfra-dev/pyinfra/issues/1085
    """

    def create_temp_dir(parent_dir: str, user: str) -> str:
        class TempDir(typing.TypedDict):  # pylint: disable=missing-class-docstring
            path: str
            created: bool
            queued: bool

        temp_dirs = vars(pyinfra_temp_path).setdefault(self.name, {})
        temp_dir: TempDir = temp_dirs.setdefault(
            user, {"path": "", "created": False, "queued": False}
        )

        # One temporary directory per host and user combination should
        # be created.
        if not temp_dir["path"]:
            temp_dir["path"] = f"{parent_dir}/cac.{uuid.uuid4().hex[:10]}"

        # Temporary directories should only be created when needed
        # during the execution stage.
        if pyinfra.state.is_executing:
            if not temp_dir["created"]:
                # A fact is used to execute commands so that operation
                # parameters are automatically inherited.
                self.get_fact(
                    pyinfra.facts.server.Command,
                    (
                        shlex.join(["mkdir", "-m", "700", "--", temp_dir["path"]])
                        + " && "
                        + shlex.join(["chown", "--", user, temp_dir["path"]])
                    ),
                )
                temp_dir["created"] = True
        elif not self.in_op:
            # If this function is not called as part of an operation,
            # an operation has to be queued to ensure that the directory
            # creation code is executed during the execution stage.
            if not temp_dir["queued"]:
                pyinfra.operations.python.call(
                    name="Create temporary directory: " + temp_dir["path"],
                    function=create_temp_dir,
                    parent_dir=parent_dir,
                    user=user,
                )
                temp_dir["queued"] = True

        return temp_dir["path"]

    # SFTP uploads are executed with the user specified in the connector.
    # If an operation uses privilege escalation, any uploads end up in
    # the temporary directory and are moved to the desired location with
    # escalated privileges. By using a separate temporary directory for
    # this process, the connected user does not have to receive access
    # to the superuser's temporary directory.
    user = None
    if isinstance(self.connector, pyinfra.connectors.ssh.SSHConnector):
        # Using stack frame objects to check whether this function is
        # being called as part of the put operation is simpler than
        # monkey patching operation functions.
        # pylint: disable-next=protected-access
        caller_frame = sys._getframe().f_back
        if caller_frame:
            if (
                caller_frame.f_code.co_filename.endswith("/pyinfra/operations/files.py")
                and caller_frame.f_code.co_name == "put"
            ):
                user = self.get_fact(
                    pyinfra.facts.server.User, _sudo=False, _su_user=None, _doas=False
                )
    if user is None:
        user = self.get_fact(pyinfra.facts.server.User)

    # Most arguments passed to the function are ignored because it
    # should always return a random path to prevent collisions.
    return create_temp_dir(temp_directory or "/tmp", user) + "/" + uuid.uuid4().hex[:14]


pyinfra.api.host.Host.get_temp_filename = pyinfra_temp_path  # type: ignore
```

With this approach, pyinfra creates a temporary directory with a random name and permission bits set to 700 (`/tmp` is used statically because I don't require `$TMPDIR` evaluation). All temporary file/directory paths generated by pyinfra are scoped to this directory. The temporary directory is created when needed during the execution stage to prevent needless creation of temporary directories. If an SFTP upload is executed with an unprivileged user, an additional temporary directory for that user is created. In addition, most arguments passed to `pyinfra.api.host.Host.get_temp_filename()` are ignored, because the returned path should always be random to prevent collisions.

**Comment by sudoBash418** [comment]
I'd love to see this implemented, as it can a pretty big security hole in my opinion (depending on use-case).

FWIW, I [tried implementing](https://git.sb418.net/sudoBash418/pyinfra/commits/branch/private-temp-directories) a similar feature against pyinfra 2.4 a couple years ago, but didn't have time to polish it for upstreaming - I might pick it back up and post a PR.

One notable issue I ran into was that using `sudo`/`su`/`doas` with non-privileged target users tends to be incompatible with this feature. This is because basic Unix permissions don't allow you to assign read permissions to an arbitrary user (root can usually bypass FS access controls, but other users usually cannot).
~~[POSIX ACLs](https://wiki.archlinux.org/title/Access_Control_Lists) might be a solution to this, but `setfacl` isn't guaranteed to exist and neither is filesystem support. I suppose pyinfra could just fallback to either raising an error, or setting files world-readable, depending on a config flag? That way there's an escape hatch for weird situations, but pyinfra would become more secure-by-default.~~

Edit: somehow I missed the fact that `setfacl` [is already used](https://github.com/pyinfra-dev/pyinfra/blob/3.x/pyinfra/connectors/ssh.py#L483) whenever a user is specified :smile:

**Comment by TobyGiacometti** [comment]
Good point about using unprivileged users with `sudo`/`su`/`doas`! The workaround I shared does not account for this as I don't need it, but it wouldn't complicate things too much: Multiple temporary directories for each SFTP user and `sudo`/`su`/`doas` user combination could be created, each with an ACL entry for the `sudo`/`su`/`doas` user. If ACLs are not supported, I agree that raising an error would be the safest solution, while at the same time offering an opt-in solution to still create world-readable files (for affected user combinations). As such files are moved right after upload, potentially sensitive information is not world-readable for long.

**Comment by Chaoses-Ib** [comment]
For `pyinfra.operations.server.script()`, how about just piping the script from stdin? This would be more secure and faster. I'm not sure why pyinfra chose to write a temp file instead. Is it bacause there are shells only supports file input but not stdin, or to avoid reading stdin in scripts accidentally?

**Comment by wowi42** [comment]
I reviewed this against the current 3.x code and the report is still accurate. Posting a concrete proposal based on what is in the tree today.

## Current state

- `Host.get_temp_filename` (`src/pyinfra/api/host.py:335`) still generates flat paths in `config.TEMP_DIR` (default `/tmp`) as `pyinfra-<sha1(hash_key)>`. When callers pass a deterministic `hash_key`, the name is predictable.
- `apt.deb` passes the download URL as the hash key (`src/pyinfra/operations/apt.py:423`) and calls `files.download._inner` without `force` or a checksum. A local unprivileged user on the target can pre-create a package at the predictable path and have it installed. This is the most severe part of the issue.
- `server.script` (`src/pyinfra/operations/server.py:244`) uploads via SFTP as the connected user with the default umask, then `chmod +x`. On a typical system that leaves the script world-readable for the duration of the run.

## Proposed solution

Two independent changes, in order of priority:

**1. Fix `apt.deb` (and similar download consumers) first.** Pass `force=True` or a checksum to `files.download` so a pre-existing file at the temp path is never trusted. This is a small, self-contained fix that closes the worst hole on its own and can land as its own PR.

**2. Introduce a per-host secure temp directory.** Shape it roughly like the workaround in the issue body, but implemented in core:

- At execution start, lazily create one directory per host and effective user under `TEMP_DIR` with a random name and mode 0700, owned by the connected user.
- `get_temp_filename` returns random paths inside that directory. The `hash_key` parameter stops influencing the on-disk name; it should be kept only for cache/idempotence semantics if any caller depends on it.
- Privilege escalation handover: the SSH connector already runs `setfacl -m u:<user>:r` on uploaded temp files when `_sudo_user`/`_su_user`/`_doas_user`/`_dzdo_user` is set (`src/pyinfra/connectors/ssh.py:612`). The secure temp directory needs the same treatment: grant the escalation user an ACL on the directory itself (at minimum `x`, plus `r` where needed), not just the file.
- Fallback behavior when `setfacl` is missing or the filesystem has no ACL support (minimal Alpine/BSD installs): fail loudly by default when an escalation user differs from the connected user, with a config flag to opt back into world-readable temp files for those environments.
- Non-SSH connectors (local, docker, chroot) need the same handling or a documented opt-out.

The stdin piping suggestion avoids the temp file for `server.script` specifically, but it changes semantics (scripts that read stdin would break) and does nothing for `files.put`, `apt.deb`, or `gpg`, which genuinely need a remote file. I would not pursue it as the general fix.

Happy to work on a PR for step 1 if that split sounds reasonable.
